libpgcommon/lwgeom_cache.c

```diff
@@ -236,4 +239,64 @@ GetGeomCache(FunctionCallInfo fcinfo,
 	return NULL;
 }
 
+/******************************************************************************/
 
+inline static ToastCache*
+ToastCacheGet(FunctionCallInfo fcinfo)
+{
+	const uint32_t entry_number = TOAST_CACHE_ENTRY;
+	GenericCacheCollection* generic_cache = GetGenericCacheCollection(fcinfo);
+	ToastCache* cache = (ToastCache*)(generic_cache->entry[entry_number]);
+	if (!cache)
+	{
+		cache = MemoryContextAllocZero(FIContext(fcinfo), sizeof(ToastCache));
+		cache->type = entry_number;
+		generic_cache->entry[entry_number] = (GenericCache*)cache;
+	}
+	return cache;
+}
+
+GSERIALIZED*
+ToastCacheGetGeometry(FunctionCallInfo fcinfo, uint32_t argnum)
+{
+	Assert(argnum < ToastCacheSize);
+	ToastCache* cache = ToastCacheGet(fcinfo);
+	ToastCacheArgument* arg = &(cache->arg[argnum]);
+
+	Datum datum = PG_GETARG_DATUM(argnum);
+	struct varlena *attr = (struct varlena *) DatumGetPointer(datum);
+
+	/* Argument is not TOASTED, access as normal. */
+	if (!VARATT_IS_EXTERNAL_ONDISK(attr))
+		return (GSERIALIZED*)PG_DETOAST_DATUM(datum);
+
+	/* Retrieve the unique keys */
+	struct varatt_external ve;
+	VARATT_EXTERNAL_GET_POINTER(ve, attr);
+	Oid valueid = ve.va_valueid;
+	Oid toastrelid = ve.va_toastrelid;
+
+	/* We've seen this object before? */
+	if (arg->valueid == valueid && arg->toastrelid == toastrelid)
+	{
+		if (arg->geom)
+			return arg->geom;
+
+		/* Take a copy into the upper context */
+		MemoryContext old_context = MemoryContextSwitchTo(FIContext(fcinfo));
+		arg->geom = (GSERIALIZED*)PG_DETOAST_DATUM_COPY(datum);
+		MemoryContextSwitchTo(old_context);
+		return arg->geom;
+	}
+	/* New object, clear our old copies and see if it */
```

Comment:
Is it worth avoiding the extra memcpy in the first hit for the case where all values are new?

Reply:
I go back and forth on that, but I think that (a) in the case where the objects are repeating, we'll get more than 2-3 repetitions and the extra cost will wash out and (b) in the case where the objects are not repeating, we don't want to incur the extra copy and free cost of putting a copy into the upper context.